File: app/src/main/java/com/spoongecko/app/SecureCredentialManager.java
package com.spoongecko.app;

import android.content.Context;
import android.content.SharedPreferences;

import androidx.security.crypto.EncryptedSharedPreferences;
import androidx.security.crypto.MasterKey;

import org.json.JSONArray;
import org.json.JSONObject;

import java.io.BufferedReader;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;
import java.util.Locale;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

final class SecureCredentialManager {

    private static final String PREFS_FILE = "spoon_secure_credentials";
    private static final String KEY_INDEX = "__vault_index__";
    private static final String SUFFIX_USER = "_user";
    private static final String SUFFIX_PASS = "_pass";
    private static final String SUFFIX_PRIMARY = "_primary_user";

    private static SecureCredentialManager instance;
    private final SharedPreferences prefs;
    private final ExecutorService io = Executors.newSingleThreadExecutor();

    static final class Entry {
        public final String host;
        public final String username;
        public final String password;

        Entry(String host, String username, String password) {
            this.host = host;
            this.username = username;
            this.password = password;
        }
    }

    interface CredentialsCallback {
        void onResult(List<Entry> entries);
    }

    interface BooleanCallback {
        void onResult(boolean value);
    }

    interface StringCallback {
        void onResult(String value);
    }

    private static final class Credential {
        final String host;
        final String username;
        final String password;

        Credential(String host, String username, String password) {
            this.host = host;
            this.username = username;
            this.password = password;
        }
    }

    private SecureCredentialManager(Context context) {
        try {
            MasterKey masterKey = new MasterKey.Builder(context)
                    .setKeyScheme(MasterKey.KeyScheme.AES256_GCM)
                    .build();
            prefs = EncryptedSharedPreferences.create(
                    context,
                    PREFS_FILE,
                    masterKey,
                    EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
                    EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM);
        } catch (Exception e) {
            throw new IllegalStateException("Failed to initialize encrypted vault", e);
        }
    }

    static synchronized SecureCredentialManager get(Context context) {
        if (instance == null) {
            instance = new SecureCredentialManager(context.getApplicationContext());
        }
        return instance;
    }

    void saveCredentials(String rawHost, String username, String password) {
        String host = sanitizeHost(rawHost);
        String user = username == null ? "" : username.trim();
        if (host.isEmpty() || user.isEmpty() || password == null || password.isEmpty()) return;
        io.execute(() -> store(host, user, password));
    }

    void hasCredential(String rawHost, String username, String password, BooleanCallback callback) {
        String host = sanitizeHost(rawHost);
        String user = username == null ? "" : username.trim();
        if (host.isEmpty() || password == null) {
            callback.onResult(false);
            return;
        }
        io.execute(() -> {
            boolean same = false;
            synchronized (prefs) {
                String stored = prefs.getString(keyPass(host, user), null);
                if (stored != null && stored.equals(password)) {
                    same = true;
                }
                if (!same) {
                    String primary = prefs.getString(keyPrimary(host), null);
                    if (primary != null) {
                        String primaryPass = prefs.getString(keyPass(host, primary), null);
                        if (primaryPass != null && primaryPass.equals(password)) {
                            same = true;
                        }
                    }
                }
            }
            callback.onResult(same);
        });
    }

    void getAllCredentials(CredentialsCallback callback) {
        io.execute(() -> {
            List<Entry> out = new ArrayList<>();
            synchronized (prefs) {
                for (Credential credential : readIndex()) {
                    out.add(new Entry(credential.host, credential.username, credential.password));
                }
            }
            callback.onResult(out);
        });
    }

    void deleteCredentials(String rawHost, String username) {
        String host = sanitizeHost(rawHost);
        String user = username == null ? "" : username.trim();
        if (host.isEmpty() || user.isEmpty()) return;
        io.execute(() -> {
            synchronized (prefs) {
                prefs.edit()
                        .remove(keyUser(host, user))
                        .remove(keyPass(host, user))
                        .apply();
                String primary = prefs.getString(keyPrimary(host), null);
                if (user.equals(primary)) {
                    String replacement = null;
                    for (Credential credential : readIndex()) {
                        if (credential.host.equals(host) && !credential.username.equals(user)) {
                            replacement = credential.username;
                            break;
                        }
                    }
                    if (replacement == null) {
                        prefs.edit().remove(keyPrimary(host)).apply();
                    } else {
                        prefs.edit().putString(keyPrimary(host), replacement).apply();
                    }
                }
                List<Credential> index = readIndex();
                for (int i = index.size() - 1; i >= 0; i--) {
                    Credential credential = index.get(i);
                    if (credential.host.equals(host) && credential.username.equals(user)) {
                        index.remove(i);
                    }
                }
                writeIndex(index);
            }
        });
    }

    void importFromCsv(InputStream input, Runnable done) {
        io.execute(() -> {
            importFromCsvSync(input);
            if (done != null) done.run();
        });
    }

    void getExportCsv(StringCallback callback) {
        io.execute(() -> callback.onResult(exportCsvSync()));
    }

    private void store(String host, String username, String password) {
        synchronized (prefs) {
            prefs.edit()
                    .putString(keyUser(host, username), username)
                    .putString(keyPass(host, username), password)
                    .putString(keyPrimary(host), username)
                    .apply();
            List<Credential> index = readIndex();
            boolean found = false;
            for (Credential credential : index) {
                if (credential.host.equals(host) && credential.username.equals(username)) {
                    found = true;
                    break;
                }
            }
            if (!found) {
                index.add(new Credential(host, username, password));
            }
            writeIndex(index);
        }
    }

    private void importFromCsvSync(InputStream input) {
        try (BufferedReader reader = new BufferedReader(
                new InputStreamReader(input, StandardCharsets.UTF_8))) {
            String header = reader.readLine();
            if (header == null) return;
            String[] columns = splitCsv(header);
            int colUrl = -1;
            int colUser = -1;
            int colPass = -1;
            for (int i = 0; i < columns.length; i++) {
                String column = columns[i].trim().toLowerCase(Locale.ROOT);
                if (colUrl < 0 && (column.equals("url") || column.equals("host")
                        || column.equals("website") || column.equals("site"))) {
                    colUrl = i;
                } else if (colUser < 0 && (column.equals("username")
                        || column.equals("user") || column.equals("login"))) {
                    colUser = i;
                } else if (colPass < 0 && (column.equals("password")
                        || column.equals("pass"))) {
                    colPass = i;
                }
            }
            if (colUser < 0 || colPass < 0) return;
            String line;
            while ((line = reader.readLine()) != null) {
                if (line.trim().isEmpty()) continue;
                String[] parts = splitCsv(line);
                if (parts.length <= Math.max(colUser, colPass)) continue;
                String host = colUrl >= 0 && colUrl < parts.length ? parts[colUrl] : "";
                String user = parts[colUser];
                String pass = parts[colPass];
                String normalized = sanitizeHost(host);
                if (normalized.isEmpty()) normalized = "imported";
                String normalizedUser = user == null ? "" : user.trim();
                if (normalizedUser.isEmpty() || pass == null || pass.isEmpty()) continue;
                store(normalized, normalizedUser, pass);
            }
        } catch (Exception ignored) {
        }
    }

    private String exportCsvSync() {
        synchronized (prefs) {
            StringBuilder sb = new StringBuilder("host,username,password\n");
            for (Credential credential : readIndex()) {
                sb.append(csvField(credential.host)).append(',')
                        .append(csvField(credential.username)).append(',')
                        .append(csvField(credential.password)).append('\n');
            }
            return sb.toString();
        }
    }

    private List<Credential> readIndex() {
        List<Credential> result = new ArrayList<>();
        String json = prefs.getString(KEY_INDEX, "[]");
        try {
            JSONArray array = new JSONArray(json);
            for (int i = 0; i < array.length(); i++) {
                JSONObject object = array.getJSONObject(i);
                String host = object.optString("host", "");
                String username = object.optString("username", "");
                if (host.isEmpty() || username.isEmpty()) continue;
                String password = prefs.getString(keyPass(host, username), null);
                if (password != null) {
                    result.add(new Credential(host, username, password));
                }
            }
        } catch (Exception ignored) {
        }
        return result;
    }

    private void writeIndex(List<Credential> index) {
        JSONArray array = new JSONArray();
        for (Credential credential : index) {
            JSONObject object = new JSONObject();
            try {
                object.put("host", credential.host);
                object.put("username", credential.username);
            } catch (Exception ignored) {
                continue;
            }
            array.put(object);
        }
        prefs.edit().putString(KEY_INDEX, array.toString()).apply();
    }

    private String keyUser(String host, String username) {
        return host + "_" + username + SUFFIX_USER;
    }

    private String keyPass(String host, String username) {
        return host + "_" + username + SUFFIX_PASS;
    }

    private String keyPrimary(String host) {
        return host + SUFFIX_PRIMARY;
    }

    private static String sanitizeHost(String host) {
        if (host == null) return "";
        String value = host.trim().toLowerCase(Locale.ROOT);
        int scheme = value.indexOf("://");
        if (scheme >= 0) value = value.substring(scheme + 3);
        int slash = value.indexOf('/');
        if (slash >= 0) value = value.substring(0, slash);
        int at = value.indexOf('@');
        if (at >= 0) value = value.substring(at + 1);
        int colon = value.lastIndexOf(':');
        if (colon > 0 && value.indexOf(']') < colon) value = value.substring(0, colon);
        value = value.trim();
        while (value.startsWith(".")) value = value.substring(1);
        return value;
    }

    private static String csvField(String value) {
        if (value == null) return "";
        if (value.contains(",") || value.contains("\"") || value.contains("\n")) {
            return "\"" + value.replace("\"", "\"\"") + "\"";
        }
        return value;
    }

    private static String[] splitCsv(String line) {
        List<String> out = new ArrayList<>();
        StringBuilder current = new StringBuilder();
        boolean inQuotes = false;
        for (int i = 0; i < line.length(); i++) {
            char ch = line.charAt(i);
            if (inQuotes) {
                if (ch == '"') {
                    if (i + 1 < line.length() && line.charAt(i + 1) == '"') {
                        current.append('"');
                        i++;
                    } else {
                        inQuotes = false;
                    }
                } else {
                    current.append(ch);
                }
            } else {
                if (ch == '"') {
                    inQuotes = true;
                } else if (ch == ',') {
                    out.add(current.toString());
                    current.setLength(0);
                } else {
                    current.append(ch);
                }
            }
        }
        out.add(current.toString());
        return out.toArray(new String[0]);
    }
}

File: app/src/main/java/com/spoongecko/app/SecureCredentialManager.kt
package com.spoongecko.app

import android.content.Context
import android.content.SharedPreferences
import android.net.Uri
import android.widget.Toast
import androidx.security.crypto.EncryptedSharedPreferences
import androidx.security.crypto.MasterKey
import org.json.JSONArray
import org.json.JSONObject
import org.mozilla.geckoview.Autocomplete
import java.io.BufferedReader
import java.io.InputStreamReader
import java.io.OutputStreamWriter

class SecureCredentialManager(context: Context) {
    private var encryptedPrefs: SharedPreferences? = null
    private var isReady = false

    init {
        try {
            val masterKey = MasterKey.Builder(context.applicationContext)
                .setKeyScheme(MasterKey.KeyScheme.AES256_GCM)
                .build()
            encryptedPrefs = EncryptedSharedPreferences.create(
                context.applicationContext,
                "spoon_secure_vault",
                masterKey,
                EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
                EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM
            )
            isReady = true
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    @Synchronized
    fun saveCredentials(host: String, username: String, password: String) {
        if (!isReady || host.isEmpty() || username.isEmpty()) return
        val cleanUser = username.trim()
        encryptedPrefs?.edit()
            ?.putString("${host}_${cleanUser}_user", cleanUser)
            ?.putString("${host}_${cleanUser}_pass", password)
            ?.putString("${host}_primary_user", cleanUser)
            ?.apply()
    }

    @Synchronized
    fun getUsername(host: String): String {
        if (!isReady) return ""
        return encryptedPrefs?.getString("${host}_primary_user", "") ?: ""
    }

    @Synchronized
    fun getPassword(host: String): String {
        if (!isReady) return ""
        val username = getUsername(host)
        if (username.isEmpty()) return ""
        return encryptedPrefs?.getString("${host}_${username}_pass", "") ?: ""
    }

    @Synchronized
    fun getAllAccountsForHost(host: String): String {
        if (!isReady || host.isEmpty()) return "[]"
        return try {
            val array = JSONArray()
            val all = encryptedPrefs?.all ?: return "[]"
            for (key in all.keys) {
                if (key.startsWith("${host}_") && key.endsWith("_user") && !key.endsWith("_primary_user")) {
                    val username = all[key] as? String ?: continue
                    if (username.isEmpty()) continue
                    val password = encryptedPrefs?.getString("${host}_${username}_pass", "") ?: ""
                    val obj = JSONObject()
                    obj.put("username", username)
                    obj.put("password", password)
                    array.put(obj)
                }
            }
            array.toString()
        } catch (e: Exception) { "[]" }
    }

    @Synchronized
    fun getAllCredentialsAsJson(): String {
        val array = JSONArray()
        if (!isReady) return array.toString()
        try {
            val all = encryptedPrefs?.all ?: return array.toString()
            for (key in all.keys) {
                if (key.endsWith("_user") && !key.endsWith("_primary_user")) {
                    val username = all[key] as? String ?: continue
                    if (username.isEmpty()) continue
                    val suffix = "_${username}_user"
                    if (key.endsWith(suffix) && key.length > suffix.length) {
                        val host = key.substring(0, key.length - suffix.length)
                        val password = encryptedPrefs?.getString("${host}_${username}_pass", "") ?: ""
                        val obj = JSONObject()
                        obj.put("host", host)
                        obj.put("username", username)
                        obj.put("password", password)
                        array.put(obj)
                    }
                }
            }
        } catch (e: Exception) { }
        return array.toString()
    }

    @Synchronized
    fun deleteCredentials(host: String, username: String) {
        if (!isReady || host.isEmpty() || username.isEmpty()) return
        val editor = encryptedPrefs?.edit() ?: return
        editor.remove("${host}_${username}_pass")
        editor.remove("${host}_${username}_user")
        if (username == (encryptedPrefs?.getString("${host}_primary_user", "") ?: "")) {
            editor.remove("${host}_primary_user")
        }
        editor.commit()
    }

    @Synchronized
    fun editCredentialPassword(host: String, username: String, newPassword: String) {
        if (!isReady || host.isEmpty() || username.isEmpty()) return
        encryptedPrefs?.edit()
            ?.putString("${host}_${username.trim()}_pass", newPassword)
            ?.apply()
    }

    @Synchronized
    fun clearCredentials(host: String) {
        if (!isReady || host.isEmpty()) return
        val editor = encryptedPrefs?.edit() ?: return
        val all = encryptedPrefs?.all ?: return
        for (key in all.keys) {
            if (key.startsWith("${host}_")) {
                editor.remove(key)
            }
        }
        editor.apply()
    }

    // --- NEW: CSV EXPORT & IMPORT ---
    fun exportToCsv(uri: Uri, context: Context) {
        val jsonStr = getAllCredentialsAsJson()
        if (jsonStr == "[]") {
            Toast.makeText(context, "Vault is empty", Toast.LENGTH_SHORT).show()
            return
        }
        try {
            val arr = JSONArray(jsonStr)
            context.contentResolver.openOutputStream(uri)?.use { stream ->
                val writer = OutputStreamWriter(stream)
                writer.write("host,username,password\n")
                for (i in 0 until arr.length()) {
                    val obj = arr.getJSONObject(i)
                    val host = obj.optString("host", "").replace("\"", "\"\"")
                    val user = obj.optString("username", "").replace("\"", "\"\"")
                    val pass = obj.optString("password", "").replace("\"", "\"\"")
                    writer.write("\"$host\",\"$user\",\"$pass\"\n")
                }
                writer.flush()
            }
            Toast.makeText(context, "CSV Exported successfully", Toast.LENGTH_SHORT).show()
        } catch (e: Exception) {
            Toast.makeText(context, "Export failed: ${e.message}", Toast.LENGTH_SHORT).show()
        }
    }

    fun importFromCsv(uri: Uri, context: Context) {
        try {
            context.contentResolver.openInputStream(uri)?.use { stream ->
                val reader = BufferedReader(InputStreamReader(stream))
                reader.readLine() // Skip header
                var line = reader.readLine()
                var count = 0
                while (line != null) {
                    val parts = parseCsvLine(line)
                    if (parts.size >= 3 && parts[0].isNotEmpty()) {
                        saveCredentials(parts[0], parts[1], parts[2])
                        count++
                    }
                    line = reader.readLine()
                }
                Toast.makeText(context, "Imported $count logins", Toast.LENGTH_SHORT).show()
            }
        } catch (e: Exception) {
            Toast.makeText(context, "Import failed: ${e.message}", Toast.LENGTH_SHORT).show()
        }
    }

    private fun parseCsvLine(line: String): List<String> {
        val result = mutableListOf<String>()
        var current = StringBuilder()
        var inQuotes = false
        for (c in line) {
            if (c == '"') inQuotes = !inQuotes
            else if (c == ',' && !inQuotes) {
                result.add(current.toString())
                current.clear()
            } else current.append(c)
        }
        result.add(current.toString())
        return result
    }

    // --- NEW: AUTOFILL HELPER FOR GECKOVIEW ---
    fun getLoginsForDomain(domain: String): List<Autocomplete.LoginEntry> {
        val entries = mutableListOf<Autocomplete.LoginEntry>()
        if (!isReady) return entries
        try {
            val all = encryptedPrefs?.all ?: return entries
            val hosts = mutableSetOf<String>()
            for (key in all.keys) {
                if (key.endsWith("_primary_user")) {
                    hosts.add(key.removeSuffix("_primary_user"))
                }
            }
            for (host in hosts) {
                if (host.contains(domain, ignoreCase = true) || domain.contains(host, ignoreCase = true)) {
                    val user = getUsername(host)
                    val pass = getPassword(host)
                    if (user.isNotEmpty() && pass.isNotEmpty()) {
                        val origin = if (host.startsWith("http")) host else "https://$host"
                        entries.add(Autocomplete.LoginEntry.Builder().origin(origin).username(user).password(pass).build())
                    }
                }
            }
        } catch (_: Exception) {}
        return entries
    }
}
